### PR TITLE
Add GPU driver installation support in launcher

### PR DESCRIPTION
**Pull request by meetrajvala**
This PR adds the following changes for supporting the GPU driver installation in launcher component:

- Installation of open sourced GPU drivers using the `cos-gpu-installer`. 
- Changes in `container_runnner` to ensure that GPU device is accessible to the workload container.
- Utility functions for listing the GPU device files and remounting the installation directory.
- Image tests with required setup and validation scripts.
- Unit tests

### CODE REVIEW

File: launcher/container_runner.go
Comment (jkl73) [inline launcher/container_runner.go:167]: consider using https://pkg.go.dev/path/filepath#WalkDir to get all nvidia devices, may make the code more concise

File: launcher/launcher/main.go
Comment (jkl73) [inline launcher/launcher/main.go:167]: maybe list all supported GPUs in the error mesg

File: launcher/internal/gpu/driverinstaller.go
Comment (jkl73) [inline launcher/internal/gpu/driverinstaller.go:118]: move the error check up, also doesn't make sense to have installerImageRef in the error message

File: launcher/image/test/scripts/test_gpu_workload.sh
Comment (jkl73) [inline launcher/image/test/scripts/test_gpu_workload.sh:10]: where is this "TEST PASSED" coming from, from the workload?

File: launcher/internal/gpu/driverinstaller.go
Comment (jkl73) [inline launcher/internal/gpu/driverinstaller.go:83]: Are you sure it's installing to the right directory? I saw on line 66, it's mounting the root to /root/ in the container

File: launcher/internal/experiments/experiments.go
Comment (jkl73) [inline launcher/internal/experiments/experiments.go:18]: Have you try to allowlist the test project on the experiment? otherwise the cloudbuild test might not pass

File: launcher/image/test/scripts/test_gpu_workload.sh
Comment (meetrajvala) [reply launcher/image/test/scripts/test_gpu_workload.sh:10]: yes, it is coming from workload.

File: launcher/internal/experiments/experiments.go
Comment (meetrajvala) [reply launcher/internal/experiments/experiments.go:18]: yes, test project is in the allowlist.

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:83]: root is getting mounted to /root for installer container. we expect drivers to be installed under "/var/lib/nvidia" which is getting set by the host-dir flag where drivers will be installed on the host.

File: launcher/image/testworkloads/gpu/cuda-vector-add/Dockerfile
Comment (meetrajvala) [reply launcher/image/testworkloads/gpu/cuda-vector-add/Dockerfile:3]: With container registry to artifact registry migration, url will remain same. This existing url will continue to work after the migration.

File: launcher/launcher/main.go
Comment (meetrajvala) [reply launcher/launcher/main.go:167]: done

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:118]: done

File: launcher/internal/gpu/deviceinfo/deviceinfo.go
Comment (yawangwang) [inline launcher/internal/gpu/deviceinfo/deviceinfo.go:4]: Why not just use cos.googlesource.com/cos/tools.git/src/cmd/cos_gpu_installer/deviceinfo? You don't have to duplicate the same logics here.

File: launcher/internal/gpu/driverinstaller.go
Comment (yawangwang) [inline launcher/internal/gpu/driverinstaller.go:43]: Can we just run the command using `exec.Command` from https://pkg.go.dev/cos.googlesource.com/cos/tools.git@v0.0.0-20241008015903-8431fe581b1f/src/cmd/cos_gpu_installer#section-readme to install GPU driver?

File: launcher/internal/gpu/deviceinfo/deviceinfo.go
Comment (meetrajvala) [reply launcher/internal/gpu/deviceinfo/deviceinfo.go]: Done.

File: launcher/image/test/scripts/test_gpu_unsupported_gputype.sh
Comment (yawangwang) [inline launcher/image/test/scripts/test_gpu_unsupported_gputype.sh:10]: could you refactor your script to accept another param to check if the expected message exists in serial logs? That way you don't need three scripts with same logics.

**Comment by yawangwang** [comment]
Please squash your commits to have a cleaner history.

File: launcher/image/test/scripts/test_gpu_unsupported_gputype.sh
Comment (meetrajvala) [reply launcher/image/test/scripts/test_gpu_unsupported_gputype.sh:10]: along expected messages, there are different success and failure messages so I didn't parameterized it and kept it separate.

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:55]: Done. Added comment with README link.

File: launcher/util_test.go
Comment (meetrajvala) [reply launcher/util_test.go]: Done

File: launcher/internal/gpu/driverinstaller.go
Comment (jkl73) [inline launcher/internal/gpu/driverinstaller.go:134]: This feels weird,

can you try one thing? try to create the directory before launching the install container? And see if the result driver is executable without remounting

File: launcher/internal/gpu/driverinstaller.go
Comment (jkl73) [inline launcher/internal/gpu/driverinstaller.go:142]: can we hardcode this ref? I run the command I saw "us.gcr.io/cos-cloud/cos-gpu-installer:v2.4.1"
I assume this won't change if in the same image?

File: launcher/container_runner.go
Comment (meetrajvala) [reply launcher/container_runner.go]: Yes, specOpts contains `mounts`. Updated it to directly append `gpuMounts`.

File: launcher/util.go
Comment (meetrajvala) [reply launcher/util.go]: Done

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:142]: As we use image family (e.g. cos-113-lts`) field in cloudbuild, build will use the latest image from the family which may have the installer version updated and may break if we hardcode it.  So it is recommended to get the version using `cos-extensions` for given base cos image as `cos-gpu-installer` is not guaranteed to be backward compatible.

File: launcher/util.go
Comment (yawangwang) [inline launcher/util.go:36]: nit: use nil initialization: `var targetFiles []string`

File: launcher/util.go
Comment (yawangwang) [inline launcher/util.go:40]: The error handling logics should be carefully evaluated. From https://pkg.go.dev/io/fs#WalkDirFunc looks like simply returning a non-nil error will stop `WalkDir` walking the entire tree, is this the intended behavior?

File: launcher/internal/gpu/driverinstaller.go
Comment (yawangwang) [inline launcher/internal/gpu/driverinstaller.go:124]: cleanup task after its done.

File: launcher/internal/gpu/driverinstaller.go
Comment (yawangwang) [inline launcher/internal/gpu/driverinstaller.go:101]: cleanup container when it's done.

File: launcher/internal/gpu/driverinstaller.go
Comment (yawangwang) [inline launcher/internal/gpu/driverinstaller.go:136]: what if the returned code is not 0, do we just log the code and return nil? We should apply retry to non-zero code and return an error instead of nil.

File: launcher/launcher/main.go
Comment (yawangwang) [reply launcher/launcher/main.go]: sry but I'm still not seeing any reason blocking doing install + gpu consumption in the same place. You can possibly create a new container with a different ID from the existing workload container and do the installation from there.

File: launcher/image/test/scripts/test_gpu_unsupported_gputype.sh
Comment (yawangwang) [reply launcher/image/test/scripts/test_gpu_unsupported_gputype.sh:10]: then could you put those three separate files into a new directory something like `scripts/gpu` since they're highly related to gpu tests.

**Review by yawangwang — CHANGES_REQUESTED** [review]

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:101]: done

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:124]: done

File: launcher/launcher/main.go
Comment (meetrajvala) [reply launcher/launcher/main.go]: gpu consumption related code snippet needs to be part of `NewRunner` function of `container_runnner` as the workload container gets created (not launched) as part of this constructor method. Calling the `InstallGpuDrivers` method from this `NewRunner` would add more responsibility to this method as this method should only create the runner object with required fields.

so `InstallGpuDrivers` function should be called before we call `NewRunner` method, so kept it as part of `startLauncher`.

File: launcher/util.go
Comment (meetrajvala) [reply launcher/util.go:40]: We will skip the directories for which `err` argument is set to non-nil during the `WalkDirFunc` call.

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:136]: Make sense to log and return an error. Returned error would not be retryable as we do not want to retry on the gpu driver installation failure.

File: launcher/util.go
Comment (yawangwang) [inline launcher/util.go:38]: I guess you also need to do a `d.IsDir` check, otherwise `WalkDir` will skip the path's parent directory, resulting in missing reading potential target files.

File: launcher/launcher/main.go
Comment (yawangwang) [reply launcher/launcher/main.go]: If you view gpu installation as a preliminary work for creating a runner object, then I'm okay with your approach here.

File: launcher/launcher/main.go
Comment (yawangwang) [inline launcher/launcher/main.go:159]: Imagine a scenario when experiment is off but `InstallGpuDriver` is on, do we want to continue the launcher execution, or log the error and then exit early? I'll leave it to you to make the call.

File: launcher/util.go
Comment (meetrajvala) [reply launcher/util.go:38]: Yes, Good catch.

File: launcher/launcher/main.go
Comment (meetrajvala) [reply launcher/launcher/main.go:159]: It would good to exit early and instead of waiting for the workload to fail.

File: launcher/cloudbuild.yaml
Comment (yawangwang) [inline launcher/cloudbuild.yaml:260]: nit: create a bug item to track TODOs.

**Review by yawangwang — APPROVED** [review]
Please ensure to get approvals from other reviewers before merging this PR. Their expertise in CS image will also provide valuable insights.

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go]: If we just create the directory with required permission before installation step, then it still fails. But if we run the same `remountAsExecutable` function before the installation steps, then it works fine.

**Review by jkl73 — APPROVED** [review]

File: launcher/image/test/create_gpu_vm.sh
Comment (alexmwu) [inline launcher/image/test/create_gpu_vm.sh:2]: I don't understand why this needs to be a separate script rather than extending the existing util.

File: launcher/image/test/scripts/gpu/test_gpu_nogpu.sh
Comment (alexmwu) [inline launcher/image/test/scripts/gpu/test_gpu_nogpu.sh:17]: nit: missing newline

File: launcher/launcher/main.go
Comment (alexmwu) [inline launcher/launcher/main.go:159]: This shouldn't be done at the beginning of startLauncher. At minimum it should happen before the TPM check but probably after retrieving the auth token as well.

File: launcher/launcher/main.go
Comment (alexmwu) [inline launcher/launcher/main.go:167]: Users can't do anything about this:

"GPU installation is not enabled for this project. Please contact support."

File: launcher/launcher/main.go
Comment (alexmwu) [inline launcher/launcher/main.go:168]: in comments here and everywhere else, GPU should be capitalized

File: launcher/util.go
Comment (alexmwu) [inline launcher/util.go:41]: In what circumstances can err be non-nil and this path still be valid? It worries me that we're skipping a non-nil err check

File: launcher/internal/gpu/driverinstaller.go
Comment (alexmwu) [inline launcher/internal/gpu/driverinstaller.go:57]: nit: this additional wrapping string is likely unnecessary. You can just return err.

File: launcher/internal/gpu/driverinstaller.go
Comment (alexmwu) [inline launcher/internal/gpu/driverinstaller.go:76]: COS to be capitalized, same with GPU. For consistency, don't add a space before the colon:

"COS GPU installer version: %s", installerImageRef

File: launcher/internal/gpu/driverinstaller.go
Comment (alexmwu) [reply launcher/internal/gpu/driverinstaller.go:150]: This could be hardcoded by running this command at build time. That's because it will always be built on the same COS version. This should be fixed by GA; please add a TODO.

File: launcher/internal/gpu/driverinstaller.go
Comment (alexmwu) [inline launcher/internal/gpu/driverinstaller.go:110]: Do we need this if we mount /dev?

File: launcher/internal/gpu/driverinstaller.go
Comment (alexmwu) [inline launcher/internal/gpu/driverinstaller.go:115]: since a lot of this is duplicated, it would be useful to have helper functions that help mount with hostnamespace, hosts file and resolve conf.

File: launcher/internal/gpu/driverinstaller.go
Comment (alexmwu) [inline launcher/internal/gpu/driverinstaller.go:113]: Why do we need the same PID namespace?

**Review by alexmwu — COMMENTED** [review]
Mostly minor comments.

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:57]: Done

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:110]: Right, this can be removed as we mount `/dev`.

File: launcher/cloudbuild.yaml
Comment (meetrajvala) [reply launcher/cloudbuild.yaml]: We have to update the non confidential GPU quota to address this. As hardened image is more important, we enabled GpuDriverInstallationHardenedImageTests. GpuDriverInstallationDebugImageTests will be enabled once GPU quota is updated.

File: launcher/image/test/create_gpu_vm.sh
Comment (meetrajvala) [reply launcher/image/test/create_gpu_vm.sh:2]: We can extend the existing script. making this change as part of same PR would require updating existing tests too and increases this PR size. So I am planning to address it in follow-up PR.

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:113]: We are trying to convert [this usage command](https://cos.googlesource.com/cos/tools/+/refs/heads/master/src/cmd/cos_gpu_installer#usage) of `cos_gpu_installer` to go code and I added it as per the usage command.

File: launcher/internal/gpu/driverinstaller.go
Comment (meetrajvala) [reply launcher/internal/gpu/driverinstaller.go:115]: Will address this in follow-up PR.

**Pull request merged**
